File: day-8a/src/main.py
from __future__ import annotations

from collections.abc import Sequence, Mapping
from dataclasses import dataclass
from enum import Enum
from itertools import cycle
from typing import NewType

class StepKind(Enum):
    RIGHT = "R"
    LEFT = "L"


type StepSequence = Sequence[StepKind]
Place = NewType("Place", str)

@dataclass(frozen=True, slots=True, kw_only=True)
class Node:
    place: Place
    leftwards: Place
    rightwards: Place


type NodeMap = Mapping[Place, Node]


def move(
    *, from_: Node, direction: StepKind, node_map: NodeMap
) -> Node:
    new_place = from_.leftwards if direction is StepKind.LEFT else from_.rightwards
    return node_map[new_place]


def compute_steps_needed(puzzle_input: PuzzleInput) -> int:
    node = puzzle_input.node_map[Place("AAA")]
    steps_taken = 0
    direction_iter = cycle(puzzle_input.step_sequence)
    while node.place != Place("ZZZ"):
        node = move(
            from_=node, direction=next(direction_iter), node_map=puzzle_input.node_map
        )
        steps_taken += 1
    return steps_taken


@dataclass(frozen=True, slots=True, kw_only=True)
class PuzzleInput:
    step_sequence: StepSequence
    node_map: NodeMap


def parse_input(filename: str) -> PuzzleInput:
    with open(filename) as f:
        unparsed_input = f.read()
    first_line, rest = unparsed_input.split("\n\n")
    steps = [StepKind(char) for char in first_line]
    node_map: dict[Place, Node] = {}
    for line in rest.splitlines():
        place, _, rest = line.partition(" = ")
        left, _, right = rest.strip("()").partition(", ")
        node_map[Place(place)] = Node(
            place=Place(place), leftwards=Place(left), rightwards=Place(right)
        )
    return PuzzleInput(step_sequence=steps, node_map=node_map)


def solve(filename: str) -> int:
    puzzle_input = parse_input(filename)
    return compute_steps_needed(puzzle_input)


def main() -> None:
    print(solve("input.txt"))


if __name__ == "__main__":
    main()

File: day-8a/src/main.rs
use std::collections::HashMap;
use std::fs::read_to_string;

#[derive(Clone, Copy)]
enum StepKind {
    Left,
    Right
}


#[derive(Clone)]
struct Node {
    place: String,
    leftwards: String,
    rightwards: String
}


fn step<'a>(
    from: &'a Node, direction: &'a StepKind, node_map: &'a HashMap<String, Node>
) -> &'a Node {
    match direction {
        StepKind::Left => &node_map[&from.leftwards],
        StepKind::Right => &node_map[&from.rightwards]
    }
}


fn compute_steps_needed(puzzle_input: PuzzleInput) -> u32 {
    let mut node = puzzle_input.node_map.get("AAA").unwrap();
    let mut steps_taken = 0;
    let mut direction_iter = puzzle_input.step_sequence.iter().cycle();
    while node.place != "ZZZ" {
        let direction = direction_iter.next().unwrap();
        node = step(&node, direction, &puzzle_input.node_map);
        steps_taken += 1;
    };
    steps_taken
}


struct PuzzleInput {
    step_sequence: Vec<StepKind>,
    node_map: HashMap<String, Node>
}


fn parse_input(filename: &str) -> PuzzleInput {
    let unparsed_input = read_to_string(filename).unwrap();
    let [first_line, rest] = match unparsed_input
        .split("\r\n\r\n")
        .collect::<Vec<&str>>()[..] {
            [first, rest] => [first, rest],
            _ => panic!()
        };
    let step_sequence: Vec<StepKind> = first_line
        .chars()
        .map(|c| match c {
                'L' => StepKind::Left,
                'R' => StepKind::Right,
                _ => panic!()
            }
        )
        .collect();
    let mut node_map: HashMap<String, Node> = HashMap::new();
    for line in rest.lines() {
        let [place, rest] = match line.split(" = ").collect::<Vec<&str>>()[..] {
            [place, rest] => [place.to_string(), rest.to_string()],
            _ => panic!()
        };
        let [left, right] = match rest
            .trim_start_matches('(')
            .trim_end_matches(')')
            .split(", ")
            .collect::<Vec<&str>>()[..] {
                [left, right] => [left.to_string(), right.to_string()],
                _ => panic!()
            };
        node_map.insert(
            place.clone(), Node {place, leftwards: left, rightwards: right}
        );
    }
    PuzzleInput {step_sequence, node_map}
}


fn solve(filename: &str) -> u32 {
    let puzzle_input = parse_input(filename);
    compute_steps_needed(puzzle_input)
}

fn main() {
    println!("{}", solve("input.txt"));
}
